Update aspire hosting samples (#8265)
Particular/docs.particular.net: platform/aspire/index.md

## Viewing the platform in the Aspire dashboard

The platform appears in the [Aspire dashboard](https://aspire.dev/dashboard/) as a single `ParticularPlatform` parent resource. The components the integration creates are nested as children, each surfacing the URL exposed by its primary endpoint:

- **ServicePulse**: the web UI
- **ServiceControl Error**: the error instance API
- **ServiceControl Audit**: the audit instance API
- **ServiceControl Monitoring**: the monitoring instance API
- **Management Studio**: the RavenDB management UI (when using the managed RavenDB persistence instance)

Externally supplied resources are not nested under the platform. An Azure Service Bus or RabbitMQ broker passed in via `AddConnectionString("...")` (or any other `Add*` call), and an existing RavenDB instance attached via `WithPersistenceRavenDb(existingRaven)`, appear as separate top-level resources in the dashboard. The platform does not own their lifecycle, so it does not group them under itself.

The platform node tracks readiness as its children come up. It starts in `Starting`, transitions to `Running` once every child reports healthy, and moves to `RuntimeUnhealthy` if any child stops.

> 

Reviewer: ```suggestion
```

### Azure Service Bus

`ThroughputReportingAzureServiceBus` supplies Azure AD credentials so the Error instance can query Service Bus management APIs. Pass each value as an Aspire parameter; mark the client secret as a secret parameter.

snippet: aspire-throughput-asb

See [Usage reporting when using the Azure Service Bus transport](/servicecontrol/servicecontrol-instances/configuration.md#usage-reporting-when-using-the-azure-service-bus-transport) for what each value is used for in ServiceControl.

#### Options

| Option | Default |
| --- | --- |
| `tenantId` (`IExpressionValue`) constructor parameter | Required |
| `subscriptionId` (`IExpressionValue`) constructor parameter | Required |
| `clientId` (`IExpressionValue`) constructor parameter | Required |
| `clientSecret` (`IExpressionValue`) constructor parameter | Required, should be a secret parameter |
| `serviceBusName` (`IExpressionValue?`) constructor parameter | Inferred from the transport configuration when omitted |
| `managementUrl` (`IExpressionValue?`) constructor parameter | Public Azure management URL |

> 

Reviewer: ```suggestion
```

### RabbitMQ

`ThroughputReportingRabbitMq` supplies RabbitMQ management API credentials so the Error instance can query broker statistics. All parameters are optional; ServiceControl falls back to the broker connection string for any value not supplied.

snippet: aspire-throughput-rabbitmq

See [Usage reporting when using the RabbitMQ transport](/servicecontrol/servicecontrol-instances/configuration.md#usage-reporting-when-using-the-rabbitmq-transport) for what each value is used for in ServiceControl.

#### Options

| Option | Default |
| --- | --- |
| `apiUrl` (`IExpressionValue?`) constructor parameter | Falls back to the RabbitMQ transport connection string |
| `userName` (`IExpressionValue?`) constructor parameter | Falls back to the RabbitMQ transport connection string |
| `password` (`IExpressionValue?`) constructor parameter | Falls back to the RabbitMQ transport connection string, should be a secret parameter when supplied |

> 

Reviewer: ```suggestion
```

For production deployments, supply the `<platform-name>-license` parameter through the standard Aspire parameter input mechanisms (user secrets, environment variables, command-line arguments, or a secret-store integration) so the value never lives in the published manifest. The in-code defaults remain useful for local development.

### Configuring host ports

Each managed container exposes an endpoint allowing Aspire to manage the exposed endpoint to avoid conflicts with ports already in use by other processes. As these ports are selected randomly by Aspire at startup you can refer to the dashboard for the url to access each component.

In some cases it may be desirable to have a fixed port exposed by these components to provide a stable url (for example `http://localhost:9090` for ServicePulse) without having to inspect the Aspire dashboard.

Override a port with the standard Aspire `WithEndpoint(endpointName, callback)` API:

snippet: aspire-host-ports

> 

Reviewer: ```suggestion
```

For production deployments, supply the `<platform-name>-license` parameter through the standard Aspire parameter input mechanisms (user secrets, environment variables, command-line arguments, or a secret-store integration) so the value never lives in the published manifest. The in-code defaults remain useful for local development.

### Configuring host ports

Each managed container exposes an endpoint allowing Aspire to manage the exposed endpoint to avoid conflicts with ports already in use by other processes. As these ports are selected randomly by Aspire at startup you can refer to the dashboard for the url to access each component.

In some cases it may be desirable to have a fixed port exposed by these components to provide a stable url (for example `http://localhost:9090` for ServicePulse) without having to inspect the Aspire dashboard.

Override a port with the standard Aspire `WithEndpoint(endpointName, callback)` API:

snippet: aspire-host-ports

> Refer to the following table for the default endpoint names and ports exposed by each component:

Reviewer: ```suggestion
```

| Component                   | Endpoint name  | Internal Port  |
| --------------------------- | -------------- | -------------- |
| ServicePulse                | `servicepulse` | 9090           |
| ServiceControl Error        | `error`        | 33333          |
| ServiceControl Audit        | `audit`        | 44444          |
| ServiceControl Monitoring   | `monitoring`   | 33633          |

> | Managed RavenDB persistence | `http`         | 8080           |

Reviewer: Remove

Reviewer: ```suggestion
```

## Viewing the platform in the Aspire dashboard

The platform appears in the [Aspire dashboard](https://aspire.dev/dashboard/) as a single `ParticularPlatform` parent resource. The components the integration creates are nested as children, each surfacing the URL exposed by its primary endpoint:

- **ServicePulse**: the web UI
- **ServiceControl Error**: the error instance API
- **ServiceControl Audit**: the audit instance API
- **ServiceControl Monitoring**: the monitoring instance API
- **Management Studio**: the RavenDB management UI (when using the managed RavenDB persistence instance)

Externally supplied resources are not nested under the platform. An Azure Service Bus or RabbitMQ broker passed in via `AddConnectionString("...")` (or any other `Add*` call), and an existing RavenDB instance attached via `WithPersistenceRavenDb(existingRaven)`, appear as separate top-level resources in the dashboard. The platform does not own their lifecycle, so it does not group them under itself.

The platform node tracks readiness as its children come up. It starts in `Starting`, transitions to `Running` once every child reports healthy, and moves to `RuntimeUnhealthy` if any child stops.

> <!-- TODO: add a screenshot of the running platform in the Aspire dashboard. -->

Reviewer 2: ```suggestion
```

snippet: aspire-components-explicit

### ServiceControl Error instance

`AddServiceControlErrorInstance(name, persistence)` adds a [ServiceControl Error instance](/servicecontrol/servicecontrol-instances/), running the [`particular/servicecontrol`](https://hub.docker.com/r/particular/servicecontrol) image, as a child of the platform. The supplied persistence must have been registered on the platform via a `WithPersistence*` or `AddPersistence*` extension.

The error queue name defaults to `error`. Override it with `WithErrorQueueName(queueName)`. Use `WithThroughputQueue(queueName)` to override the queue on which throughput data is received from the monitoring instance. This is independent of throughput reporting, which separately configures broker-statistics querying.

snippet: aspire-components-error

#### Options

| Option | Default |
| --- | --- |
| `name` (`string`) parameter on `AddServiceControlErrorInstance` | Required |
| `persistence` (`IResourceBuilder<IResource>`) parameter on `AddServiceControlErrorInstance` | Required |
| `queueName` (`string`) parameter on `WithErrorQueueName` | `error` |

> | `queueName` (`string`) parameter on `WithThroughputQueue` | Not set (throughput reporting off) |

Reviewer 2: This is for receiving throughput from the monitoring instance. If no queue is specified, the default of "ServiceControl.ThroughputData" is used.
```suggestion
| `queueName` (`string`) parameter on `WithThroughputQueue` | `ServiceControl.ThroughputData` |
```

### ServiceControl Monitoring instance

`AddServiceControlMonitoringInstance(name)` adds a [ServiceControl Monitoring instance](/servicecontrol/monitoring-instances/), running the [`particular/servicecontrol-monitoring`](https://hub.docker.com/r/particular/servicecontrol-monitoring) image, as a child of the platform. The monitoring instance does not require a persistence resource.

The monitoring queue name defaults to `Particular.Monitoring`. Override it with `WithMonitoringQueueName(queueName)`. Use `WithThroughputQueueFrom(error)` to copy the throughput queue name from the error instance and keep the two aligned without repeating the name; `WithThroughputQueue(queueName)` sets it explicitly when the queue address needs to differ from the error instance.

snippet: aspire-components-monitoring

#### Options

| Option | Default |
| --- | --- |
| `name` (`string`) parameter on `AddServiceControlMonitoringInstance` | Required |
| `queueName` (`string`) parameter on `WithMonitoringQueueName` | `Particular.Monitoring` |
| `errorInstance` (`IResourceBuilder<ServiceControlErrorInstanceResource>`) parameter on `WithThroughputQueueFrom` | Required (when called) |

> | `queueName` (`string`) parameter on `WithThroughputQueue` | Not set |

Reviewer 2: Default value would be used if it's not set. 
```suggestion
| `queueName` (`string`) parameter on `WithThroughputQueue` | `ServiceControl.ThroughputData` |
```

### Azure Service Bus

`ThroughputReportingAzureServiceBus` supplies Azure AD credentials so the Error instance can query Service Bus management APIs. Pass each value as an Aspire parameter; mark the client secret as a secret parameter.

snippet: aspire-throughput-asb

See [Usage reporting when using the Azure Service Bus transport](/servicecontrol/servicecontrol-instances/configuration.md#usage-reporting-when-using-the-azure-service-bus-transport) for what each value is used for in ServiceControl.

#### Options

| Option | Default |
| --- | --- |
| `tenantId` (`IExpressionValue`) constructor parameter | Required |
| `subscriptionId` (`IExpressionValue`) constructor parameter | Required |
| `clientId` (`IExpressionValue`) constructor parameter | Required |
| `clientSecret` (`IExpressionValue`) constructor parameter | Required, should be a secret parameter |
| `serviceBusName` (`IExpressionValue?`) constructor parameter | Inferred from the transport configuration when omitted |

> | `managementUrl` (`IExpressionValue?`) constructor parameter | Public Azure management URL |

Reviewer 2: if omitted, the default is used based on some logic:https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl.Transports.ASBS/AzureQuery.cs#L112

This is why I think repeating the logic here is not ideal :-)

If you are using the Particular Service Platform with Aspire today and would like to see more Aspire support, [reach out](https://github.com/Particular/NServiceBus/issues/6941).

### Transport

| Transport                                                 | Status                       |
| --------------------------------------------------------- | ---------------------------- |
| [Learning](/transports/learning/)                         | Supported (Development only) |
| [Azure Service Bus](/transports/azure-service-bus/)       | Supported                    |
| [RabbitMQ](/transports/rabbitmq/)                         | Supported                    |
| [Amazon SQS](/transports/sqs/)                            | Supported                    |
| [Azure Storage Queues](/transports/azure-storage-queues/) | Not yet supported            |
| [Microsoft SQL Server](/transports/sql/)                  | Not yet supported            |
| [PostgreSQL](/transports/postgresql/)                     | Not yet supported            |
| [IBM MQ](/transports/ibmmq/)                              | Not yet supported            |

> | [MSMQ](/transports/msmq/)                                 | Not available in containers  |

Reviewer: Does this status make sense? Should we just say "Not supported?
```suggestion
| [MSMQ](/transports/msmq/)                                 | Not yet supported  |
```

Author: should it even be included in the table at all?

Reviewer 2: ```suggestion
```

Aspire supports multiple deployment targets, including [Docker Compose](https://aspire.dev/deployment/docker-compose/), Kubernetes, and Azure Container Apps. See [Aspire pipelines](https://aspire.dev/deployment/pipelines/) for the deployment model and the supported targets.

Two integration-specific points to keep in mind for any publish target:

- The license parameter default is materialized into the manifest at publish time. Supply the license through Aspire parameter inputs in production so the value does not live in the published output. See [Supplying the license in production](#production-considerations-supplying-the-license-in-production).

> - The Learning transport is blocked in publish mode unless explicitly opted in. See [Learning transport in publish mode](#production-considerations-learning-transport-in-publish-mode).

Author: This refers to a section that no longer exists
```suggestion
- The Learning transport is blocked in publish mode.
```

Reviewer: It does still exist - https://github.com/Particular/docs.particular.net/pull/8265/changes#diff-13046e09c17257a6933217945f8862a99e21a6ed8259c3d9e08773a99a02146eR395

Author: oh, I thought I removed it - which we absolutely should do.

The learning transport is not supported in production and we should not be encouraging users to publish it into real infrastructure. The ability to disable that behaviour was only included so that the tests could verify it since they leverage the publish pipeline.

Reviewer: Yep, lets remove it

---
title: Aspire

> summary: Describes how to orchestrate the Particular Platform via the Aspire AppHost

Reviewer 3: ```suggestion
summary: Describes how to orchestrate the Particular Platform via an Aspire AppHost
```

reviewed: 2026-05-21
component: Aspire.Hosting
related:
- samples/hosting/aspire
- samples/aspire/platform
- samples/aspire/platform-asb
---

> The `Particular.Aspire.Hosting.ServicePlatform` package is an [Aspire](https://aspire.dev/) hosting integration that runs the Particular Platform — the ServiceControl instances, ServicePulse, persistence, and message transport — as part of an Aspire AppHost. It is intended for developers and technical leads who run the platform locally during development and want the same AppHost to carry through to publish-mode deployments, without maintaining a separate set of infrastructure scripts.

Reviewer 3: ```suggestion
The `Particular.Aspire.Hosting.ServicePlatform` package is an [Aspire](https://aspire.dev/) hosting integration that runs the Particular Service Platform — the ServiceControl instances, ServicePulse, persistence, and message transport — as part of an Aspire AppHost. It is intended for developers and technical leads who run the platform locally during development and want the same AppHost to carry through to publish-mode deployments, without maintaining a separate set of infrastructure scripts.
```

## Overview

The `Particular.Aspire.Hosting.ServicePlatform` package extends an [Aspire](https://aspire.dev/) AppHost so it can run a complete Particular Platform alongside the rest of a distributed application.

A single `AddParticularPlatform(...)` call registers a platform resource that owns:

- [ServiceControl](/servicecontrol) error, audit, and monitoring instances
- [ServicePulse](/servicepulse)
- A managed RavenDB persistence instance, or one supplied by the AppHost

> - The configured message transport (Learning, Azure Service Bus, or RabbitMQ)

Reviewer 3: This list is missing SQS. Maybe remove the list and link to the section below.

- The platform license

Transport, persistence, and licensing are configured once on the platform resource and propagated to every component, so the containers start in the correct order. [NServiceBus endpoints](/nservicebus/endpoints/) attach with `WithParticularPlatform(...)` and pick up the same transport connection string and license without additional wiring.

snippet: aspire-apphost

## Supported components

The integration is being built out across the Particular Platform in stages. The tables below show what is currently wired through the AppHost.

If you are using the Particular Service Platform with Aspire today and would like to see more Aspire support, [reach out](https://github.com/Particular/NServiceBus/issues/6941).

> ### Transport

Reviewer 3: Each transport has a quick summary table. e.g. [AzureServiceBus](https://github.com/Particular/docs.particular.net/blob/master/transports/azure-service-bus/index.md?plain=1#L18-L30). I wonder if we should include Aspire compatibility in there and link back to this section (for Yes or No).

| --------------------------------------------------------- | ---------------------------- |
| [Learning](/transports/learning/)                         | Supported (Development only) |
| [Azure Service Bus](/transports/azure-service-bus/)       | Supported                    |
| [RabbitMQ](/transports/rabbitmq/)                         | Supported                    |
| [Amazon SQS](/transports/sqs/)                            | Supported                    |
| [Azure Storage Queues](/transports/azure-storage-queues/) | Not yet supported            |
| [Microsoft SQL Server](/transports/sql/)                  | Not yet supported            |
| [PostgreSQL](/transports/postgresql/)                     | Not yet supported            |
| [IBM MQ](/transports/ibmmq/)                              | Not yet supported            |

### Persistence

| Persistence                          | Status    |
| ------------------------------------ | --------- |
| [RavenDB](/servicecontrol/ravendb/)  | Supported |

## Prerequisites

- The [Aspire CLI](https://aspire.dev/get-started/install-cli/) and the .NET 10 SDK, used to build and run the AppHost project.

> - A container runtime such as [Docker Desktop](https://www.docker.com/products/docker-desktop/) or [Podman](https://podman.io/). The platform components are pulled from Docker Hub as `particular/servicecontrol`, `particular/servicecontrol-audit`, `particular/servicecontrol-monitoring`, `particular/servicecontrol-ravendb`, and `particular/servicepulse`.

Reviewer 3: Consider linking this to [here](https://docs.particular.net/servicecontrol/hosting#containers) instead as it contains a list of supported container runtimes.

Reviewer 2: ```suggestion
- A [container runtime](/servicecontrol/hosting#containers). The platform components are pulled from Docker Hub as `particular/servicecontrol`, `particular/servicecontrol-audit`, `particular/servicecontrol-monitoring`, `particular/servicecontrol-ravendb`, and `particular/servicepulse`.
```

- A Particular Platform license. See [Configuring the license](#configuring-the-license) for the license sources the integration accepts.

## Installation

If the solution does not yet have an Aspire AppHost project, follow the [Build your first Aspire app](https://aspire.dev/get-started/first-app/) quickstart to scaffold one.

Then add the integration package to the AppHost project:

```sh
dotnet add package Particular.Aspire.Hosting.ServicePlatform
```

Only the AppHost project needs this reference. NServiceBus endpoint projects pick up the platform license and transport connection string from environment variables injected by `WithParticularPlatform(...)`, so they do not need their own reference to this package.

## Quick start

`AddDefaultComponents()` wires up the Learning transport, a managed RavenDB persistence instance, the ServiceControl error, audit, and monitoring instances, and ServicePulse with sensible defaults:

> snippet: aspire-quick-start-1

Reviewer 3: Consider giving these snippets a name rather than a number. If someone wants to add one, or shuffle them around in the future, it will be confusing.

To attach an [NServiceBus endpoint](/nservicebus/endpoints/) so it picks up the platform's license and transport connection string, chain `WithParticularPlatform(platform)` on the endpoint's project resource:

snippet: aspire-quick-start-2

> These defaults are intended for local development. For deployment scenarios, configure an explicit transport, persistence, and license as described in [Configuring the transport](#configuring-the-transport), [Configuring persistence](#configuring-persistence), and [Configuring the license](#configuring-the-license).

Reviewer 3: It may be worth providing a sample snippet here with a full configuration just to show it.

- **Management Studio**: the RavenDB management UI (when using the managed RavenDB persistence instance)

Externally supplied resources are not nested under the platform. An Azure Service Bus or RabbitMQ broker passed in via `AddConnectionString("...")` (or any other `Add*` call), and an existing RavenDB instance attached via `WithPersistenceRavenDb(existingRaven)`, appear as separate top-level resources in the dashboard. The platform does not own their lifecycle, so it does not group them under itself.

The platform node tracks readiness as its children come up. It starts in `Starting`, transitions to `Running` once every child reports healthy, and moves to `RuntimeUnhealthy` if any child stops.

## Configuring the transport

The platform uses whichever transport is configured via a `WithTransport*` extension on the platform resource. The same transport connection string is then propagated to every ServiceControl instance and to any NServiceBus endpoint attached via `WithParticularPlatform(...)`. See [Supported components](#supported-components) for the transports currently wired through the integration.

### Learning transport

> The [Learning transport](/transports/learning/) stores messages on the local file system. By default they are persisted to `.learningtransport` under the AppHost project directory. Pass a custom path to override.

Reviewer 3: Ideally this would be a `.learningtransport` next to the solution file. Consider linking to https://docs.particular.net/transports/learning/#usage-storage-directory

Your AppHost is likely to be part of the same solution as your endpoints. If they default to the same location then the endpoints do not need any configuration of a specific folder.

Author: I think it's probably worthwhile to change our default to be solution-relative, it's still worth having the example here as we allow the path to be passed in though the extension method 

changing the default will make it more likely to "just work" if someone discovers our package without the samples

Reviewer 2: will make the change to the default location

snippet: aspire-transport-learning

The Learning transport appears in the Aspire dashboard as a `learning-transport` connection-string resource nested under the platform, holding the resolved storage path.

> [!WARNING]
> The Learning transport is intended for development only and the AppHost will error during publish flows when it is included.

#### Options

| Option | Default |
| --- | --- |
| `storagePath` (`string?`) parameter on `WithTransportLearning` | `.learningtransport` (relative to the AppHost project directory) |

### Azure Service Bus

The platform treats Azure Service Bus as an external resource, so the AppHost only needs a connection string to reach the namespace. Other ways to model the resource in the AppHost (such as `AddAzureServiceBus` with Aspire provisioning, or `AsExisting`) are described in [Set up Azure Service Bus in the AppHost](https://aspire.dev/integrations/cloud/azure/azure-service-bus/azure-service-bus-host/) in the Aspire documentation.

snippet: aspire-transport-asb

#### Options

| Option | Default |
| --- | --- |
| `azureServiceBus` (`IResourceBuilder<IResourceWithConnectionString>`) parameter on `WithTransportAzureServiceBus` | Required |

> Azure Service Bus tuning such as topic name, partitioning, hierarchy namespace, and websocket mode is configured on the connection string itself. See [Azure Service Bus transport configuration](/servicecontrol/transports.md#azure-service-bus) for the full list of connection-string options.

Reviewer 3: I'm not sure how this would work if I was working with a resource itself and not a connection string. That might be something to think about for a future version of the package.

Author: Do we want to break out the `aspire-transport-asb` into two snippets rather than the inline comment: one with transport and one with connection string?

Reviewer 2: keeping it as is for initial release

| `region`                   (`string`) parameter on `WithTransportAmazonSQS`                            | Required |
| `accessKey`                (`IExpressionValue`) parameter on `WithTransportAmazonSQS`                  | Required |
| `secretAccessKey`          (`IExpressionValue`) parameter on `WithTransportAmazonSQS`                  | Required |
| `configure`                (`Action<AmazonSQSTransportSettings`) parameter on `WithTransportAmazonSQS` | Optional |
| `QueueNamePrefix`          (`string?`) property on `AmazonSQSTransportSettings`                        | Optional |
| `TopicNamePrefix`          (`string?`) property on `AmazonSQSTransportSettings`                        | Optional |
| `S3BucketForLargeMessages` (`IExpressionValue?`) property on `AmazonSQSTransportSettings`              | Optional |
| `QueueNamePrefix`          (`string?`) property on `AmazonSQSTransportSettings`                        | Optional |

## Configuring persistence

ServiceControl uses a persistence backend to store error and audit data, retry state, and saga history. See [Supported components](#supported-components) for the persisters currently wired through the integration.

> Configure persistence on the platform resource, then pass the resulting persistence builder into the ServiceControl Error and Audit instances that need it. The Monitoring instance does not require persistence.

Reviewer 3: Should we have some statement about error and audit being able to share a persistence (which it does for the default configuation)? Should we be making (or preferrably linking off to) some guidance around this?

Reviewer 2: ```suggestion
Configure persistence on the platform resource, then pass the resulting persistence builder into the ServiceControl Error and Audit instances that need it. The Monitoring instance does not require persistence. Follow the [managing persistence guidance](/servicecontrol/ravendb/containers) when setting up for production use.
```

Aspire supports multiple deployment targets, including [Docker Compose](https://aspire.dev/deployment/docker-compose/), Kubernetes, and Azure Container Apps. See [Aspire pipelines](https://aspire.dev/deployment/pipelines/) for the deployment model and the supported targets.

Two integration-specific points to keep in mind for any publish target:

- The license parameter default is materialized into the manifest at publish time. Supply the license through Aspire parameter inputs in production so the value does not live in the published output. See [Supplying the license in production](#production-considerations-supplying-the-license-in-production).

> - The Learning transport is blocked in publish mode.

Reviewer 3: Both of these points are covered already above. It probably doesn't hurt to repeat them, just pointing it out in case it was not intentional.

### ServicePulse shows the monitoring panel as unavailable

`AddServicePulse` accepts an optional monitoring instance. When the monitoring argument is omitted, or `WithMonitoringInstance(null)` is called on the ServicePulse builder, ServicePulse runs without the monitoring panel. Pass an `IResourceBuilder<ServiceControlMonitoringInstanceResource>` to `AddServicePulse`, or call `WithMonitoringInstance(monitoring)` on the ServicePulse builder, to enable it. (`AddDefaultComponents()` always wires monitoring in automatically.)

### "Platform [name] has mismatched ServiceControl container image versions" in Aspire host console

This warning appears when the ServiceControl container images (error, audit, and monitoring instances) are configured with different version tags. For example:

snippet: aspire-components-version

 All ServiceControl components (error, audit, and monitoring) must run the same version to ensure compatibility. The warning lists each component with its configured image and tag so the mismatch is easy to identify.

> To resolve the warning, ensure all ServiceControl instances use the same image tag, or remove explicit `.WithImage()` calls to use the default latest tag, which always refers to a compatible set of images.

Reviewer: ```suggestion
To resolve the warning, ensure all ServiceControl instances use the same image tag, or remove explicit `.WithImage()` calls to use the default `latest` tag, which always refers to a compatible set of images.
```

| Transport                                                 | Status                      |
| --------------------------------------------------------- | --------------------------- |
| [Learning](/transports/learning/)                         | Supported                   |
| [Azure Service Bus](/transports/azure-service-bus/)       | Supported                   |
| [RabbitMQ](/transports/rabbitmq/)                         | Supported                   |
| [Amazon SQS](/transports/sqs/)                            | Not yet supported           |
| [Azure Storage Queues](/transports/azure-storage-queues/) | Not yet supported           |
| [Microsoft SQL Server](/transports/sql/)                  | Not yet supported           |
| [PostgreSQL](/transports/postgresql/)                     | Not yet supported           |
| [IBM MQ](/transports/ibmmq/)                              | Not yet supported           |
| [MSMQ](/transports/msmq/)                                 | Not available in containers |

### Persistence

| Persistence                          | Status    |
| ------------------------------------ | --------- |
| [RavenDB](/servicecontrol/ravendb/)  | Supported |

> ### Throughput reporting

Reviewer 2: I don't think we need this section as it's kind of part of the transport, and it's not a feature the user looks for - it's for our benefit only.

### Azure Service Bus

`ThroughputReportingAzureServiceBus` supplies Azure AD credentials so the Error instance can query Service Bus management APIs. Pass each value as an Aspire parameter; mark the client secret as a secret parameter.

```csharp
// TODO: replace with a real snippet from snippets/aspire
// var tenantId = builder.AddParameter("asb-tenant-id");
// var subscriptionId = builder.AddParameter("asb-subscription-id");
// var clientId = builder.AddParameter("asb-client-id");
// var clientSecret = builder.AddParameter("asb-client-secret", secret: true);
//
// var error = platform.AddServiceControlErrorInstance("servicecontrol", persistence)
//     .WithThroughputReporting(new ThroughputReportingAzureServiceBus(
//         tenantId: tenantId,
//         subscriptionId: subscriptionId,
//         clientId: clientId,
//         clientSecret: clientSecret));
```

See [Usage reporting when using the Azure Service Bus transport](/servicecontrol/servicecontrol-instances/configuration.md#usage-reporting-when-using-the-azure-service-bus-transport) for what each value is used for in ServiceControl.

> #### Options

Reviewer 2: Is it worth listing this here if these are the same as linked to above? Otherwise if those defaults\requirements change then we'll have to repeat the same here.

Reviewer: I think its worth keeping. These document the API specifically. If these changed on the transport, we'd need to change it in the package API anyway. The table also gives a bit more info on the type of param for the API.

| `serviceBusName` (`IExpressionValue?`) constructor parameter | Inferred from the transport configuration when omitted |
| `managementUrl` (`IExpressionValue?`) constructor parameter | Public Azure management URL |

### RabbitMQ

`ThroughputReportingRabbitMq` supplies RabbitMQ management API credentials so the Error instance can query broker statistics. All parameters are optional; ServiceControl falls back to the broker connection string for any value not supplied.

```csharp
// TODO: replace with a real snippet from snippets/aspire
// var apiUrl = builder.AddParameter("rabbit-api-url");
// var username = builder.AddParameter("rabbit-username");
// var password = builder.AddParameter("rabbit-password", secret: true);
//
// var error = platform.AddServiceControlErrorInstance("servicecontrol", persistence)
//     .WithThroughputReporting(new ThroughputReportingRabbitMq(apiUrl, username, password));
```

See [Usage reporting when using the RabbitMQ transport](/servicecontrol/servicecontrol-instances/configuration.md#usage-reporting-when-using-the-rabbitmq-transport) for what each value is used for in ServiceControl.

#### Options

> | Option | Default |

Reviewer 2: Is it worth listing this here if these are the same as linked to above? Otherwise if those defaults\requirements change then we'll have to repeat the same here.

For production deployments, supply the `<platform-name>-license` parameter through the standard Aspire parameter input mechanisms (user secrets, environment variables, command-line arguments, or a secret-store integration) so the value never lives in the published manifest. The in-code defaults remain useful for local development.

### Configuring host ports

Each managed container exposes an endpoint allowing Aspire to manage the exposed endpoint to avoid conflicts with ports already in use by other processes. As these ports are selected randomly by Aspire at startup you can refer to the dashboard for the url to access each component.

In some cases it may be desirable to have a fixed port exposed by these components to provide a stable url (for example `http://localhost:9090` for ServicePulse) without having to inspect the Aspire dashboard.

Override a port with the standard Aspire `WithEndpoint(endpointName, callback)` API:

```csharp
// TODO: replace with a real snippet from snippets/aspire
// platform.AddServicePulse("servicepulse", error, monitoring)
//     .WithEndpoint("servicepulse", e =>
//     {
//         e.Port = 9091;
//         e.TargetPort = 9090;
//     });
```

> Refer to the following table for the default endpoint names and ports exposed by each component:

Reviewer 2: Again, if we mention this info already elsewhere, should we just point to it instead of repeating it?

Reviewer: the `endpoint name` info isn't documented elsewhere. I think you need this info for `WithEndpoint(endpointName, callback)`, and those names are `internal const` in the package

Reviewer: In fact, maybe we need to make those consts public?

Reviewer 2: It is documented. This is the default setting for SC error: https://docs.particular.net/servicecontrol/servicecontrol-instances/configuration#transport-servicebuserrorqueue

We have the same info for all the instances, eg audit queue: https://docs.particular.net/servicecontrol/audit-instances/configuration#transport-servicebusauditqueue

Not sure what the `http` is for Managed RavenDB persistence?

Reviewer: I dont think these are the same things. These aren't the default queue names, they are the HTTP endpoint names given to aspire instance resource. the ravens "PrimaryEndpointName" is "http", so the user would need to know this if they want to set their own host port. This isnt documented anywhere (other than here).

Reviewer 2: Ah - that's confusing then - perhaps because of the use of "Endpoint name" which to me implied NSB endpoint.

Reviewer: You've worked for Particular too long ;-)

Reviewer: ```suggestion
Refer to the following table for the default HTTP endpoint names and ports exposed by each component:
```

Reviewer 2: ```suggestion
Refer to the following table for the Aspire instance resource names and ports exposed by each component:
```

Reviewer: Remove this line

`AddServicePulse` accepts an optional monitoring instance. When the monitoring argument is omitted, or `WithMonitoringInstance(null)` is called on the ServicePulse builder, ServicePulse runs without the monitoring panel. Pass an `IResourceBuilder<ServiceControlMonitoringInstanceResource>` to `AddServicePulse`, or call `WithMonitoringInstance(monitoring)` on the ServicePulse builder, to enable it. (`AddDefaultComponents()` always wires monitoring in automatically.)

### "Platform [name] has mismatched ServiceControl container image versions" in Aspire host console

This warning appears when the ServiceControl container images (error, audit, and monitoring instances) are configured with different version tags. For example:

snippet: aspire-components-version

>  All ServiceControl components (error, audit, and monitoring) must run the same version to ensure compatibility. The warning lists each component with its configured image and tag so the mismatch is easy to identify.

Reviewer: Should we link "/servicecontrol/upgrades/support-policy" somewhere ?

Reviewer 2: Maybe "must" is too strong?

```suggestion
All ServiceControl components (error, audit, and monitoring) should run the same [supported version](https://docs.particular.net/servicecontrol/upgrades/supported-versions) to ensure compatibility. The warning lists each component with its configured image and tag so the mismatch is easy to identify.
```
